Introduce deprecation policy to developer guide.

vyasr opened the pull request:
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please mark your pull request as Draft.
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove it from "Draft" and make it "Ready for Review".
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review

   If assistance is required to complete the functionality, for example when the
   C/C++ code of a feature is complete but Python bindings are still required,
   then add the label `help wanted` so that others can triage and assist.
   The additional changes then can be implemented on top of the same PR.
   If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on the target branch, force push, or rewrite history.
   Doing any of these causes the context of any comments made by reviewers to be lost.
   If conflicts occur against the target branch they should be resolved by
   merging the target branch into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->
This PR resolves #10159.

vyasr commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -745,6 +745,17 @@ void isolated_helper_function(...);
+- Deprecations should be indicated using the [deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute.
+- Deprecated APIs should be documented using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated) command.
+
+PRs containing deprecations should be labeled as "non-breaking".
+The subsequent pull request in which deprecated code is moved should be labeled as "breaking".
```
I don't know if we need to document this and can remove if we think it's extraneous. I figured it would be easier to have it here to start in case it merits discussion.

bdice replied:
I think the "breaking" policy is valuable to include. It is helpful to know and would be unclear without this note.

vyasr commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -745,6 +745,17 @@ void isolated_helper_function(...);
+If changes to libcudf involve breaking existing APIs, libcudf follows the following policy:
+- Code should be deprecated for one release prior to removal.
+- Deprecations should be indicated using the [deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute.
+- Deprecated APIs should be documented using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated) command.
```
This technically should go in DOCUMENTATION.md, not here, but I figured it was a single line and would be more discoverable here. I can add something to DOCUMENTATION.md in addition (or instead) and link here if reviewers think there would be a benefit.

harrism replied:
I think it should go in both. Please also add an example of using the doxygen `deprecated` command to DOCUMENTATION.md

vyasr replied:
Done. I also added it to an API locally and built the docs to verify that it worked as expected. Conveniently, in addition to showing up on the documentation for a particular API it looks like doxygen will generate a page containing all deprecations:

![image](https://user-images.githubusercontent.com/1538165/154392371-cea3166d-8f6d-4314-a425-a4e7116f6362.png)


![image](https://user-images.githubusercontent.com/1538165/154392324-4a201c76-d626-4e5f-8cee-c04ac2ad4cd5.png)

vuule commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -745,6 +745,17 @@ void isolated_helper_function(...);
 
+# Deprecating and Removing Code
+
+If changes to libcudf involve breaking existing APIs, libcudf follows the following policy:
```
Maybe there's more to say about this. One kind of breaking change is when we remove an API, and this text explains well what should be done in that case. More often, we replace an API, and it's not clear from this text at which point the new API should be added.
Also might be good to mention that the new API should be referenced in the deprecation message.

vyasr replied:
I've rewritten this to add some more information. Let me know what you think.

vuule reviewed (approved):
👍 for the PR. 
The promise that we'll always keep old APIs for another release is a tad scary and maybe worth an additional discussion (unless it happened and I missed it).

vyasr commented:
> +1 for the PR. The promise that we'll always keep old APIs for another release is a tad scary and maybe worth an additional discussion (unless it happened and I missed it).

Sounds good. I'm happy to get more opinions on this, perhaps in the meeting next week? I do think that promising users that their code won't break without warning is a large part of what makes a deprecation policy valuable, but I definitely don't mean to impose such a policy unilaterally.

bdice commented:
> The promise that we'll always keep old APIs for another release is a tad scary and maybe worth an additional discussion

It seems to me that libcudf's willingness to break APIs quickly (and fix them internally for our Python/Java APIs) is important for the library's development, and changing this policy could impose significant maintenance costs that we currently do not have to pay. While I believe deprecations could be important for external C++ consumers (excluding Python/Java), I have not seen explicit external requests for us to offer more warning before breaking libcudf APIs. This is a key point that I think deserves input from @jrhemstad.

jrhemstad commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -745,6 +745,22 @@ void isolated_helper_function(...);
+Changes to function parameters should be accompanied by the introduction of suitable overloads to avoid API breaks for one release cycle.
+Where possible, deprecated APIs should call new APIs, but duplication in deprecated code is acceptable at the discretion of PR reviewers if avoiding it imposes an undue burden on developers.
+
+PRs containing deprecations should be labeled as "non-breaking".
```
This is where the whole "breaking" vs. "non-breaking" thing starts to "break down" :) 

Using a `[[deprecated]]` keyword on an API will cause a compiler warning by default. If a user has warnings as errors enabled, their build is broken until they disable that warning or update to the new API. Was that a breaking change?

vyasr replied:
I chose to ignore this issue in what I wrote up. Feel free to suggest an alternative if you have a suggestion on what might be better to write here. I think using `[[deprecated]]` is still better than not, since at least for people not compiling with `Wall` we'll be offering a gentler off ramp. For people who are, the situation is no different than an outright removal.

jrhemstad reviewed (changes requested):
I'm going to suggest a radically more laissez-faire policy.

> libcudf is constantly growing and improving to improve performance and better meet our users' needs. As a result, we occasionally need to break or entirely remove APIs to respond to new and improved understanding of the functionality we provide. Remaining free to do this is essential to making libcudf an agile library that can rapidly accommodate our users needs. 

> In order to advertise these breaking changes, any PR that breaks or removes an existing API should be tagged as "breaking". This will ensure that the "Breaking" section of the release notes will include a description of what has broken from the past release. 

> On a best effort basis, the libcudf team will notify users of changes that we expect to have significant or widespread effects.

vyasr commented:
@jrhemstad I would also be fine with that policy. There are definitely pros and cons to consider here. I'll defer to whatever decision you, @harrism, @GregoryKimball, and @JohnZed feel is appropriate here, I would simply like us to have a clearly documented and agreed upon policy. I have less experience with our user base than others so it's difficult for me to weigh the tradeoffs.

jrhemstad commented:
> @jrhemstad I would also be fine with that policy. There are definitely pros and cons to consider here. I'll defer to whatever decision you, @harrism, @GregoryKimball, and @JohnZed feel is appropriate here, I would simply like us to have a clearly documented and agreed upon policy. I have less experience with our user base than others so it's difficult for me to weigh the tradeoffs.

Couldn't agree more. It's important to write down what we do even if what we write down is "We reserve the right to break your code at any time without warning", which isn't that far from the truth :)

harrism commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -745,6 +745,22 @@ void isolated_helper_function(...);
+Deprecations should be indicated using the [deprecated](https://en.cppreference.com/w/cpp/language/attributes/deprecated) attribute.
+Deprecated APIs should be documented using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated) command.
+
+When an API is deprecated, if a replacement exists the replacement should be mentioned in both the deprecation message and the deprecation documentation.
+Changes to function parameters should be accompanied by the introduction of suitable overloads to avoid API breaks for one release cycle.
+Where possible, deprecated APIs should call new APIs, but duplication in deprecated code is acceptable at the discretion of PR reviewers if avoiding it imposes an undue burden on developers.
```
Please adjust your word wrapping -- suggest keeping it to the same as our .cpp files -- 100 characters.

vyasr replied:
Done.

harrism commented:
> @jrhemstad I would also be fine with that policy. There are definitely pros and cons to consider here. I'll defer to whatever decision you, @harrism, @GregoryKimball, and @JohnZed feel is appropriate here, I would simply like us to have a clearly documented and agreed upon policy. I have less experience with our user base than others so it's difficult for me to weigh the tradeoffs.

Jake's proposed policy is basically what we follow today. I am fine with keeping that agile approach, but documenting it. libcudf has the advantage that most users interface with it via Python or Java. As long as we ensure we don't break those users (by adapting / updating bindings in the same release that we change an API), changing the C++ API is usually OK.

There are also other RAPIDS libraries that depend on libcudf, so we should be careful to not break those as well.

harrism commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -747,18 +747,25 @@ void isolated_helper_function(...);
+deprecations should also be labeled "deprecation" to facilitate discovery and removal in the
+subsequent release.
+
+In order to advertise breaking changes, any PR that breaks or removes an existing API should be
+tagged as "breaking". This will ensure that the "Breaking" section of the release notes will include
+a description of what has broken from the past release. Note that PRs containing deprecations should
+be labeled as "non-breaking".
```
```suggestion
Advertise breaking changes by labeling any pull request that breaks or removes an existing API with the "breaking" tag. This ensures that the "Breaking" section of the release notes includes
a description of what has broken from the past release. Label pull requests that contain deprecations with the "non-breaking" tag.
```

harrism commented on cpp/docs/DOCUMENTATION.md:
````diff
@@ -369,6 +369,19 @@ Although using 3 backtick marks `` ``` `` for example blocks will work too, they
+### Deprecations
+
+Add a single [@deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated) comment line
+to comment blocks for APIs that will be removed in future releases.
````
```suggestion
to comment blocks for APIs that will be removed in future releases. Mention alternative / replacement APIs in the deprecation comment.
```

harrism commented on cpp/docs/DEVELOPER_GUIDE.md:
```diff
@@ -747,25 +747,20 @@ void isolated_helper_function(...);
-tagged as "breaking". This will ensure that the "Breaking" section of the release notes will include
-a description of what has broken from the past release. Note that PRs containing deprecations should
-be labeled as "non-breaking".
+document them using Doxygen's [deprecated](https://www.doxygen.nl/manual/commands.html#cmddeprecated)
```
Line length. I know this is annoying and manual, but the rest of the doc is wrapped at the same column. Wish clang-tidy would do markdown files. This doesn't mention anything about line length: https://github.com/notslang/tidy-markdown

vyasr replied:
I think you looked at the specific commit where I pushed your changes and missed the commit afterwards where reformatted all line lengths. Can you take a look at the current version and make sure the line lengths look like you'd expect?

harrism commented:
@JohnZed would like to review before merging, but has not had time. Let's wait for him.

vyasr commented:
rerun tests

vyasr commented:
This PR is waiting for @JohnZed to take a look after GTC is completed.

ttnghia commented:
What?

JohnZed commented:
I think the text around best effort deprecation is appropriate for libcudf. It has far fewer direct callers than, say, the Python API, so we can afford to evolve it more quickly. I would like to see a more formal deprecation policy for the Python side that does ensure we do a 1-2 release deprecation plan. That should not need to block this PR. Within Python, it is often easier to provide a deprecation path, and we have far more callers 

I think the suggestion to live at head is somewhat unnecessary from this deprecation policy and likely overly prescriptive for end users. The linked video is an hour and a half, and it would be difficult if we imply users need to watch a 90 minute video to understand our deprecation approach. I don't think we're ready to meet all of the criteria that e.g. Abseil has in place for live at head. E.g. they commit to shipping refactoring tools to automate upgrades to support API removals and describe a process of "ship a tool, wait for people to apply the tool, then if practical remove the old API." This is much stricter than what is proposed elsewhere in the doc. In the absence of such tools and guarantees of easy upgrades, I suggest just omitting the "live at head" from text and keeping the best effort plan, while really ensuring we make that effort to simplify upgrades when possible.

We can revisit the best effort deprecation in the future if concrete user needs pop up. To a large extent, I think the Python API is the best place for us to have a "buffer" to allow us to upgrade libcudf APIs without impacting users of the most public-facing API, so we are just leaning on that buffer with this policy.

vyasr commented:
I have removed the recommendation (and link) to live at head. I agree that it's not explicitly necessary.

Regarding Python, I introduced a deprecation policy of 1 release when I wrote the current draft of the Python developer guide, and we have been following that rule (consistently but somewhat unofficially) for the last 3-4 releases. I am currently working on porting that developer guide into a more visible space, at which point it should become less unofficial, but I agree that the deprecation policy is more critical there and so I prioritized getting that in place first.

vyasr commented:
@gpucibot merge
